tracing-appender/src/rolling/builder.rs

```diff
@@ -0,0 +1,147 @@
+/// be passed into the constructor of `RollingFileAppenderBuilder`.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct RollingFileAppenderBuilder {
+    log_directory: String,
+    log_filename_prefix: String,
+    rotation: Option<Rotation>,
+    #[cfg(feature = "compression_gzip")]
+    compression: Option<CompressionConfig>,
+}
+
+impl RollingFileAppenderBuilder {
+    /// Creates a new `RollingFileAppnderBuilder`
+    ///
+    /// It was introduced to open up the possibility to use `compression` without
+    /// breaking the current interface.
+    ///
+    /// Note that `compression` module is enabled by using an optional feature flag
+    /// `compression_gzip` (for gzip algorithm)
+    ///
+    /// # Examples
+    /// ```rust
+    /// use tracing_appender::builder::RollingFileAppenderBuilder;
+    /// use tracing_appender::compression::CompressionOption;
+    /// use tracing_appender::rolling::Rotation;
+    /// let builder = RollingFileAppenderBuilder::new("/var/tmp", "my-app")
+    ///     .rotation(Rotation::DAILY)
+    ///     .compression(CompressionOption::GzipFast);
+    /// ```
+    pub fn new(log_directory: impl AsRef<Path>, log_filename_prefix: impl AsRef<Path>) -> Self {
+        let log_directory = log_directory
+            .as_ref()
+            .to_str()
+            .expect("Cannot convert log_directory Path to str")
+            .to_string();
+        let log_filename_prefix = log_filename_prefix
+            .as_ref()
+            .to_str()
+            .expect("Cannot convert log_filename_prefix Path to str")
+            .to_string();
+        RollingFileAppenderBuilder {
+            log_directory,
+            log_filename_prefix,
+            rotation: None,
+            #[cfg(feature = "compression_gzip")]
+            #[cfg_attr(docsrs, doc(cfg(feature = "compression_gzip")))]
+            compression: None,
+        }
+    }
+
+    /// Sets Rotation
+    pub fn rotation(mut self, rotation: Rotation) -> Self {
+        self.rotation = Some(rotation);
+        self
+    }
+
+    /// Sets compression level
+    #[cfg(feature = "compression_gzip")]
+    #[cfg_attr(docsrs, doc(cfg(feature = "compression_gzip")))]
+    pub fn compression(mut self, compression: CompressionOption) -> Self {
+        self.compression = Some(compression.into());
+        self
+    }
+
+    pub(crate) fn get_extension(&self) -> Option<String> {
+        #[cfg(feature = "compression_gzip")]
+        if let Some(compression) = self.compression.clone() {
+            compression.extension().map(|v| v.to_string())
+        } else {
+            None
+        }
+
+        #[cfg(not(feature = "compression_gzip"))]
+        None
+    }
+
+    /// Builds an instance of `RollingFileAppender` using previously defined attributes.
+    pub fn build(self) -> RollingFileAppender {
+        let now = OffsetDateTime::now_utc();
+        let rotation = self.rotation.clone().unwrap_or(Rotation::NEVER).clone();
```

Comment:
as far as i can tell, we shouldn't have to clone this twice --- does this work:
```suggestion
        let rotation = self.rotation.cloned().unwrap_or(Rotation::NEVER);
```